Sources/Plasma/PubUtilLib/plAudio/plVoiceChat.cpp

```diff
@@ -139,251 +167,227 @@ void plVoiceRecorder::SetQuality(int quality)
+                fGraph->AddData((int32_t)((100.f * sendRate) / 8.f));
+                fGraph->SetTitle(ST::format("{.2f} KiB/s", sendRate).c_str());
+            }
+            fCaptureOpenSecs = now;
         }
     }
 }
 
 plVoicePlayer::plVoicePlayer()
+    : fSound(), fOpusDecoder(plVoiceDecoder::CreateOpus())
 {
 }
 
 plVoicePlayer::~plVoicePlayer()
 {
 }
 
-void plVoicePlayer::PlaybackUncompressedVoiceMessage(void* data, unsigned size)
-{   
-    if(fEnabled)
-    {
-        if(!fSound.IsPlaying())
-        {
-            fSound.Play();      
-        }
+void plVoicePlayer::PlaybackUncompressedVoiceMessage(const void* data, size_t size, uint32_t rate)
+{
+    if (fEnabled) {
+        fSound.SetSampleRate(rate);
+        if (!fSound.IsPlaying())
+            fSound.Play();
         fSound.AddVoiceData(data, size);
     }
 }
 
-void plVoicePlayer::PlaybackVoiceMessage(void* data, unsigned size, int numFramesInBuffer)
+void plVoicePlayer::PlaybackVoiceMessage(const void* data, size_t size, int numFramesInBuffer, uint8_t flags)
 {
-    if(fEnabled)
-    {
-        int numBytes;               // the number of bytes that speex decompressed the data to. 
-        int bufferSize = numFramesInBuffer * plSpeex::GetInstance()->GetFrameSize();
-        short *nBuff = new short[bufferSize];
-        memset(nBuff, 0, bufferSize);
-
-        // Decode the encoded voice data using speex
-        if(!plSpeex::GetInstance()->Decode((uint8_t *)data, size, numFramesInBuffer, &numBytes, nBuff))
-        {
-            delete[] nBuff;
-            return;
+    if (flags & plVoiceFlags::kEncoded) {
+        plVoiceDecoder* decoder = GetDecoder(flags);
+        if (decoder) {
+            int numBytes;
+            int bufferSize = numFramesInBuffer * decoder->GetFrameSize();
+            std::unique_ptr<short> nBuff{ new short[bufferSize] };
+            if (decoder->Decode(data, size, numFramesInBuffer, numBytes, nBuff.get()))
+                PlaybackUncompressedVoiceMessage(nBuff.get(), numBytes, decoder->GetSampleRate());
         }
-        
-        uint8_t* newBuff;
-        newBuff = (uint8_t*)nBuff;         // Convert to uint8_t data
-        PlaybackUncompressedVoiceMessage(newBuff, numBytes);    // playback uncompressed data
-        delete[] nBuff;
+    } else {
+        PlaybackUncompressedVoiceMessage(data, size, FREQUENCY);
     }
 }
 
 void plVoicePlayer::SetVelocity(const hsVector3 vel)
 {
     fSound.SetVelocity(vel);
 }
-     
+
 void plVoicePlayer::SetPosition(const hsPoint3 pos)
 {
     fSound.SetPosition(pos);
 }
-    
+
 void plVoicePlayer::SetOrientation(const hsPoint3 pos)
 {
     fSound.SetConeOrientation(pos.fX, pos.fY, pos.fZ);
 }
 
+plVoiceDecoder* plVoicePlayer::GetDecoder(uint8_t voiceFlags) const
+{
+    if (voiceFlags & plVoiceFlags::kEncodedOpus)
+        return fOpusDecoder;
+    if (!(voiceFlags & plVoiceFlags::kEncoded))
+        return nullptr;
+    return plVoiceDecoder::GetSpeex();
```

Comment:
You have a flag specifically for Speex, why not check it, and leave `nullptr` as the default/fallback case?

Reply:
The Cyan standard client sends speex data flagged only as `kEncoded`. I wanted to be able to handle this data gracefully.